proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,532 @@
+  RevokeOldSecrets  bool `json:"revokeOldSecrets"`
+}
+
+type OIDCClientSecretRequestStatus struct {
+  GeneratedSecret    string `json:"generatedSecret,omitempty"`
+  TotalClientSecrets int    `json:"totalClientSecrets"`
+}
+```
+
+Unlike token credential request, OIDC client secret request will require that `metadata.name` be set (so that it can
+determine what OAuth client is being referred to).  When `.spec.generateNewSecret` is set to `true`, the response will
+provide the plaintext client secret via the `.status.generatedSecret` field.  This is the only time that the plaintext
+client secret is made available.  To aid in rotation, this API may be called multiple times with
+`.spec.generateNewSecret` set to `true` to cause the creation of a new client secret.  The response will include the
+total number of client secrets (including any newly generated ones) that exist for the OAuth client in the
+`.status.totalClientSecrets` field.  When the admin is ready, they may call the API with `.spec.revokeOldSecrets` set to
+`true` to cause all but the latest secret to be revoked.  In the event of a client secret disclosure, a "hard" rotation
+may be performed by setting both `.spec.generateNewSecret` and `.spec.revokeOldSecrets` to `true` (this will revoke all
+pre-existing client secrets and return a newly generated secret).  Leaving both of these fields set to `false` will
+simply return the number of existing client secrets via the `.status.totalClientSecrets` field (this same information is
+available via the `.status.totalClientSecrets` field of the `OIDCClient` resource).
+
+A dynamic client with many client secrets will be slower to authenticate (especially slow when failing to authenticate),
+since each hashed secret must be tried in sequence until one works, and the cost of trying each is relatively high.
+Administrators should take care to remember to come back to use this endpoint again to revoke the old client secrets
+once they are not needed anymore.
+
+An admin would interact with this API by using standard `kubectl` commands:
+
+```yaml
+apiVersion: oauth.virtual.supervisor.pinniped.dev/v1alpha1  # different group to avoid collision with the CRD
+kind: OIDCClientSecretRequest
+metadata:
+  name: client.oauth.pinniped.dev-j77kz
+  namespace: pinniped-supervisor
+spec:
+  generateNewSecret: true
+```
+
+Assuming the above yaml is stored in `file.yaml`, then running:
+
+`kubectl create -f file.yaml`
+
+would cause the server to respond with (note the custom columns in the table output):
+
+```
+NAMESPACE               NAME                                SECRET     TOTAL
+pinniped-supervisor     client.oauth.pinniped.dev-j77kz     12..xz     2
+```
+
+All client secret hashes for an OAuth client will be stored in `pinniped-storage-oidc-client-secret-<metadata.uid>` where
+`metadata.uid` is a hashed version of the UID generated by the Kuberentes API server for the `OIDCClient` custom resource instance that
```

Comment:
```suggestion
`metadata.uid` is a base32 encoded version of the UID generated by the Kuberentes API server for the `OIDCClient` custom resource instance that
```

Reply:
Updated in 7450fb6c8e9937e43fd5f0f95f2caa674413e3eb.